courses/machine_learning/deepdive/09_sequence_keras/translate/utils_preproc.py

```diff
@@ -0,0 +1,79 @@
+"""Utilities for preprocessing natural language for machine learning"""
+
+import unicodedata
+import re
+
+import tensorflow as tf
+
+def unicode_to_ascii(s):
+        return ''.join(c for c in unicodedata.normalize('NFD', s)
+            if unicodedata.category(c) != 'Mn')
+
+def preprocess_sentence(w):
+    w = unicode_to_ascii(w.lower().strip())
+
+    # creating a space between a word and the punctuation following it
+    # eg: "he is a boy." => "he is a boy ."
+
+    w = re.sub(r"([?.!,¿])", r" \1 ", w)
+    w = re.sub(r'[" "]+', " ", w)
+
+    # replacing everything with space except (a-z, A-Z, ".", "?", "!", ",")
+    w = re.sub(r"[^a-zA-Z?.!,¿]+", " ", w)
```

Comment:
* Shouldn't that line go before w = re.sub(r'[" "]+', " ", w)? My reasoning is that you first replace whatever is not a char or a final punctuation sign by a space, and then you remove the extra spaces created. Not the reverse.